README.markdown
Loot-o-matic addon for Elder Scrolls Online
===========================================

Loot-o-matic is an addon for Elder Scrolls online that helps you manage the
loot that you obtain throughout your journey.

# Features

- Filters that allow you to mark items as junk Example
  - Item Type is Trash
  - Item Type is Lure
  - Item Trait is Ornate Jewelry
  - Item ID is X
  - Item sellPrice is greater than X gold
  - Possibilities are endless!
- Filters can be enabled or disabled
- Each filter can have multiple rules
  - Item Type is X and Item Quality is X
  - Item Type is Armor and Item Trait is Ornate
- Filter types: EqualTo, GreaterThan, LessThan, Contains, NotEqualTo,
  GreaterThanOrEqualTo, LessThanOrEqualTo, DoesNotContain
- Auto marks Trash items as junk
- Auto sell junk items when opening a vendor store

# Manual Install

@TODO

# Usage

Currently this script is console based and can only be used in a chat window.

Commands:

    /lootomatic config <command>
    /lootomatic filters <command>

## config list

Used to display the current settings.

    /lootomatic config list

## config setting value

Used to change the configuration settings

Examples:

    /lootomatic config sellalljunk true
    /lootomatic config sellalljunk false
    /lootomatic config loglevel 0
    /lootomatic config loglevel 100
    /lootomatic config loglevel 200
    /lootomatic config loglevel 300

sellalljunk = Auto sell junk when vendor window opens

loglevel = Verbosity of output by Lootomatic. 0 is off, 100 is DEBUG, 200 is INFO, 300 is WARN

## filters list

This command will list all of your current loot filters.

    /lootomatic filters list

## filters show

Shows more detail about a filter.

    /lootomatic filters show 1

## filters clear

Clear all loot filters.

    /lootomatic filters clear

## filters add

Add a new loot filter

    /lootomatic filters add name:Trash enabled:true condition.type:EqualTo condition.name:itemType condition.value:48

## filters modify/update

@TODO

## filters delete

Deletes a filter by index

    /lootomatic filters delete 1

# Conditions

When you add a filter, you are defining a list of rules. Each filter will need
to have at least one rule to be able to function. The following rule types are
supported.

* EqualTo
* GreaterThan
* LessThan
* Contains
* NotEqualTo
* GreaterThanOrEqualTo
* LessThanOrEqualTo
* DoesNotContain

Along with the condition you will need to give it a name equal to that
of the property on the Item. (There's a list below) You will also
need to specify a value.

# Item Data

The following list is what you will need to set as a name when adding a rule
to your filter.

* id
* name
* itemType
* itemStyle
* sellPrice
* meetsUsageRequirement
* equipType
* itemStyle

More will be coming soon

# Support

If you are having trouble with this addon, please submit a support request
on the ESOUI website.

* http://www.esoui.com/portal.php?id=114&a=listbugs

# Feature Requests

If you are wanting to have a feature adding to this addon, please submit
the feature request on the ESOUI website.

* http://www.esoui.com/portal.php?id=114&a=listfeatures

# Development

This addon follows Semantic Versioning and it also follows the gitflow
work flow of development. If either of those sound new to you, please read
the following articles:

* http://semver.org
* http://nvie.com/posts/a-successful-git-branching-model

# Contribute

https://github.com/JoshuaEstes/eso-lootomatic

# Packaging

To package this addon for distribution the following command needs to be ran:

    git archive --format=zip --prefix=Lootomatic/ HEAD > Lootomatic.zip

This will create a new zip file in the current directory ready to be uploaded
to an addon site.

# License

Copyright (C) 2014 Joshua Estes <Joshua@Estes.in>

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG
0.4.0
	- Removed all unused code
	- Added simple rules engine for filters
	- Updated the slash commands
	- Refactored directory structure
	- Ability to remove and modify filters

0.3.0
	- Fix for quest items
	- Code clean up
	- Cleaned up some more debug messages
	- Setup configuration system
	- Made filters work a little better

0.2.0
	- Added the ability to add filters based on itemType 
	- Added the ability to list filters
	- Added the ability to show details on filters
	- Added a simple logger

0.1.0
	- Initial release with mostly debug output, not much functionality


Lootomatic/Item.lua
--[[
-- Represents an item in ESO
--
-- Item Schema
-- {
--     name                  = string,
--     color                 = string,
--     id                    = number,
--     icon                  = string,
--     sellPrice             = number,
--     meetsUsageRequirement = boolean,
--     equipType             = number,
--     itemStyle             = number,
--     itemType              = number
-- }
--
--]]

LootomaticItem = {}
LootomaticItem.__index = LootomaticItem

--[[
-- @param string itemName
--
-- @return LootomaticItem
--]]
function LootomaticItem.New(itemName)
    local self = setmetatable({data={}}, LootomaticItem)
    -- Not sure what all of these are
    self.name,  self.color, self.n3,  self.id,
    self.n5,  self.n6, self.n7,  self.n8,
    self.n9,  self.n10, self.n11, self.n12,
    self.n13, self.n14, self.n15, self.n16,
    self.n17, self.n18, self.n19, self.n20,
    self.n21, self.n22, self.n23, self.n24 = ZO_LinkHandler_ParseLink(itemName)
    self.icon, self.sellPrice, self.meetsUsageRequirement, self.equipType, self.itemStyle = GetItemLinkInfo(itemName)
    return self
end

--[[
-- Loads an item based on where it is based on the slot in a bag
--
-- @param integer bagId
-- @param integer slotId
-- @return LootomaticItem
--]]
function LootomaticItem.LoadByBagAndSlot(bagId, slotId)
    local i = LootomaticItem.New(GetItemLink(bagId, slotId, LINK_STYLE_DEFAULT))
    i.itemType = GetItemType(bagId, slotId)
    return i
end


lib/Ruler/Ruler.lua
--[[
--
-- This is a port of a generic rules engine framework
-- @see https://github.com/bobthecow/Ruler
--
--]]

Ruler = {}
Ruler.Rule = {}
Ruler.Rule.__index = Ruler.Rule

--[[
-- @param table condition
-- @return Rule
--]]
function Ruler.Rule.New(condition)
    local self = setmetatable({['condition'] = condition}, Ruler.Rule)
    return self
end

--[[
-- @param table context
-- @return boolean
--]]
function Ruler.Rule:Evaluate(context)
    return self.condition:Evaluate(context)
end

Ruler.Value = {}
Ruler.Value.__index = Ruler.Value

--[[
-- @param mixed value
-- @return Value
--]]
function Ruler.Value.New(value)
    local self = setmetatable({['value'] = value}, Ruler.Value)
    return self
end

--[[
-- @return mixed
--]]
function Ruler.Value:GetValue()
    return self.value
end

--[[
-- @param Value value
-- @return boolean
--]]
function Ruler.Value:EqualTo(value)
    return (self.value == value:GetValue())
end

--[[
-- @param Value value
-- @return boolean
--]]
function Ruler.Value:Contains(value)
end

--[[
-- @param Value value
-- @return boolean
--]]
function Ruler.Value:GreaterThan(value)
    return (self.value > value:GetValue())
end

--[[
-- @param Value value
-- @return boolean
--]]
function Ruler.Value:LessThan(value)
    return (self.value < value:GetValue())
end

Ruler.Variable = Ruler.Variable or {}
Ruler.Variable.__index = Ruler.Variable

--[[
-- @param string name
-- @param mixed value
-- @return Variable
--]]
function Ruler.Variable.New(name, value)
    local self = setmetatable({['name'] = name, ['value'] = value}, Ruler.Variable)
    return self
end

--[[
-- @return string
--]]
function Ruler.Variable:GetName()
    return self.name
end

--[[
-- @return mixed
--]]
function Ruler.Variable:GetValue()
    return self.value
end

--[[
-- @param table context
-- @return Value
--]]
function Ruler.Variable:PrepareValue(context)
    if nil ~= self.name and nil ~= context[self.name] then
        value = context[self.name]
    else
        value = self.value
    end

    return Ruler.Value.New(value)
end

Ruler.Operator = {}
Ruler.Operator.ComparisonOperator = {}
Ruler.Operator.ComparisonOperator.__index = Ruler.Operator.ComparisonOperator

--[[
-- @param Variable
-- @param Variable
-- @return ComparisonOperator
--]]
function Ruler.Operator.ComparisonOperator.New(left, right)
    local self = setmetatable({['left'] = left, ['right'] = right}, Ruler.Operator.ComparisonOperator)
    return self
end

Ruler.Operator.EqualTo = {}
Ruler.Operator.EqualTo.__index = Ruler.Operator.EqualTo
function Ruler.Operator.EqualTo.New(left, right)
    local self = setmetatable({['left'] = left, ['right'] = right}, Ruler.Operator.EqualTo)
    return self
end

--[[
-- @param table context
-- @return boolean
--]]
function Ruler.Operator.EqualTo:Evaluate(context)
    return (self.left:PrepareValue(context):EqualTo(self.right:PrepareValue(context)))
end


Lootomatic/Filter.lua
--[[
--
-- Filter Schema
-- {
--     name    = 'Display Name',
--     enabled = true,
--     rules   = { Condition, ... }
-- }
--
-- Condition Schema
-- {
--     type  = 'EqualTo',
--     left  = { Variable },
--     right = { Variable }
-- }
--
-- Variable Schame
-- {
--     name  = 'itemType',
--     value = 'value'
-- }
--
--]]

LootomaticFilter = {}
LootomaticFilter.__index = LootomaticFilter

--[[
-- @param table defaults
--
-- @return Lootomatic.Filter
--]]
function LootomaticFilter.New(defaults)
    local self = setmetatable(defaults, LootomaticFilter)
    return self
end

--[[
-- @return string
--]]
function LootomaticFilter:GetName()
    return self.name
end

--[[
-- @return boolean
--]]
function LootomaticFilter:IsEnabled()
    return self.enabled
end

--[[
-- @return table
--]]
function LootomaticFilter:GetRules()
    return self.rules
end

--[[
-- Checks to see if item matches a filter
--
-- @param LootItem lootItem
--
-- @return boolean
--]]
function LootomaticFilter:IsMatch(lootItem)
    if not self:IsEnabled() then
        return false
    end
    for i,v in pairs(self:GetRules()) do
        local o
        local c = v.condition
        --Lootomatic.Log('['..i..'] ' .. c.left.name .. ' is ' .. c.type .. ' ' .. c.left.value, LootomaticLogger.INFO)
        local var1 = Ruler.Variable.New('filter.' .. c.left.name, c.left.value)
        local var2 = Ruler.Variable.New(c.left.name)
        if 'EqualTo' == c.type then
            o = Ruler.Operator.EqualTo.New(var1, var2)
        end
        if nil ~= 0 then
            local r = Ruler.Rule.New(o)
            local e = r:Evaluate(lootItem)
            return e
        end
    end

    return false
end


Lootomatic/Settings.lua
--[[
--
-- Creates the Lootomatic Settings window.
--
--]]
Lootomatic_Settings = {}
Lootomatic_Settings.__index = Lootomatic_Settings
local LAM = LibStub( 'LibAddonMenu-1.0' )

--[[
-- @param table db
--]]
function Lootomatic_Settings.New(db)
    local self = setmetatable(
        {
            db    = db,
            panel = nil
        },
        Lootomatic_Settings
    )
    self:_Construct()
    return self
end

--[[
-- Returns the value of a setting
--
-- @param string name
-- @return mixed
--]]
function Lootomatic_Settings:GetSetting(name)
    return self.db.config[name];
end

--[[
-- Set a setting
--
-- @param string name
-- @param mixed value
--]]
function Lootomatic_Settings:SetSetting(name, value)
    self.db.config[name] = value
end

--[[
-- This toggles boolean settings
--
-- @param string name
--]]
function Lootomatic_Settings:ToggleSetting(name)
    local s = self.db.config[name]
    if false == s then
        self.db.config[name] = true
    else
        self.db.config[name] = false
    end
end

--[[
-- Configures the settings window
--]]
function Lootomatic_Settings:_Construct()
	self.panel = LAM:CreateControlPanel("Lootomatic_Settings_Panel", "Lootomatic")
    self:_AddHeader()
    self:_AddSellJunkToggle()
    self:_AddLoggerSlider()
end

--[[
-- Add the section header
--]]
function Lootomatic_Settings:_AddHeader()
	LAM:AddHeader(self.panel, "Lootomatic_Header_Settings", "Settings")
end

--[[
-- Adds checkbox to settings window that allows players to toggle if they
-- want this addon to sell junk when a vendor window opens
--]]
function Lootomatic_Settings:_AddSellJunkToggle()
    LAM:AddCheckbox(
        self.panel,
        'Lootomatic_Settings_SellAllJunk',
        'Sell All Junk?',
        'When you open a vendor store, if enabled, this will sell all items marked as junk.',
        function ()
            return self:GetSetting('sellAllJunk')
        end,
        function ()
            self:ToggleSetting('sellAllJunk')
        end,
        true,
        'Setting this to enabled WILL SELL ALL ITEMS marked as junk when you open a vendors store.'
    )
end

--[[
-- Adds a slider that allows players to set the verbosity of output
-- by Lootomatic
--]]
function Lootomatic_Settings:_AddLoggerSlider()
    LAM:AddSlider(
        self.panel,
        'Lootomatic_Settings_Logger',
        'Log Level',
        'Log level, set to 0 to disable all output. 100 will output DEBUG, INFO, and WARN messages; 200 will output INFO and WARN messages. 300 only outputs WARN messages.',
        0,
        300,
        100,
        function ()
            return self:GetSetting('logLevel')
        end,
        function (level)
            self.db.config.logLevel = level
        end
    )
end


LootomaticInterface.lua
--[[
--
-- Lootomatic GUI Interface
--
-- /-------------------\
-- | Name              |
-- | Filter Rules      |
-- | [update] [delete] |
-- |-------------------|
-- | Name              |
-- | Filter Rules      |
-- | [update] [delete] |
-- |-------------------|
-- \-------------------/
--
--]]

local wm = GetWindowManager()
LootomaticGUI = {}

--[[
-- @param string controlName
-- @param 
-- @param ControlType controlType
-- @return 
--]]
function LootomaticGUI:CreateControl(controlName, parent, controlType)
    return wm:CreateControl(controlName, parent, controlType)
end

--[[
-- @param string controlName
-- @return 
--]]
function LootomaticGUI:CreateTopLevelWindow(controlName)
    return wm:CreateTopLevelWindow(controlName)
end

--[[
-- @param string controlName
-- @param 
-- @param string virtualName
-- @return
--]]
function LootomaticGUI:CreateControlFromVirtual(controlName, parent, virtualName)
    return wm:CreateControlFromVirtual(controlName, parent, virtualName)
end

function LootomaticGUI:CreateWindow()
end

function LootomaticGUI:CreateButton()
end

function LootomaticGUI:CreateHeader()
end


Lootomatic.lua
--[[
--
-- Lootomatic
--
-- Manages loot by allowing you to set filters on what you
-- want to keep and what you want to mark as junk to sell
-- at your friendly vendor.
--
--]]
local Lootomatic   = {}
Lootomatic.name    = 'Lootomatic'
Lootomatic.version = 2

-- Defaults
Lootomatic.defaults = {
    config = {
        logLevel    = 200, -- Default is to only show INFO messages
        sellAllJunk = true
    },
    filters = {
        {
            name = 'Trash',
            enabled = true,
            markAsJunk = true,
            loot = true,
            rules = {
                {
                    condition = {
                        type = 'EqualTo',
                        left = { name = 'itemType', value = 48 },
                        right = { name = '', value = '' }
                    }
                }
            }
        }
    }
}

Lootomatic.operators = {
    EqualTo = {},
    GreaterThan = {},
    LessThan = {},
    Contains = {},
    NotEqualTo = {},
    GreaterThanOrEqualTo = {},
    LessThanOrEqualTo = {},
    DoesNotContain = {}
}

Lootomatic.variables = {
    -- DAMAGE_TYPE_
    -- EQUIP_TYPE_
    -- ITEM_QUALITY_
    -- ITEM_TRAIT_TYPE_
    -- ITEMTYPE_
    itemType = {}
    -- WEAPONTYPE_
}

-- Container for slash commands
-- @TODO move to own class
LootomaticCommands = {}

-- Used for logging output to console
-- @TODO move to own class
LootomaticLogger = {
    DEBUG = 100,
    INFO  = 200,
    WARN  = 300,
    levels = { [100] = 'DEBUG', [200] = 'INFO', [300] = 'WARN' }
}

--[[
-- @TODO Move to util class
-- @param mixed v
-- @return boolean
--]]
function toboolean(v)
    if (type(v) == 'string') then
        if ('yes' == v or 'y' == v or 'true' == v) then
            return true
        else
            return false
        end
    end

    if (type(v) == 'number') then
        if 0 == v then
            return false
        else
            return true
        end
    end

    if (type(v) == 'boolean') then
        return v
    end

    error('Cannot convert value to boolean')
end

--[[
-- Displays text in the chat window
--
-- @param string  text
-- @param integer level
--]]
function Lootomatic.Log(text, level)
    local logLevel = Lootomatic.db.config.logLevel
    if nil == level then
        level = LootomaticLogger.INFO
    end

    -- logger is disabled
    if 0 == logLevel then
        return
    end

    if nil ~= logLevel and level >= logLevel then
        d('[Lootomatic] ' .. LootomaticLogger.levels[level] .. ' ' .. tostring(text))
        return
    end
end

--[[
-- @param integer eventCode
--]]
function Lootomatic.OnOpenStore(eventCode)
    Lootomatic.Log('onOpenStore', LootomaticLogger.DEBUG)
    if Lootomatic.db.config.sellAllJunk then
        SellAllJunk()
        Lootomatic.Log('All junk items sold', LootomaticLogger.INFO)
    end
end

--[[
-- @param integer eventCode
-- @param integer bagId
-- @param integer slotId
-- @param boolean isNewItem
-- @param integer itemSoundCategory
-- @param integer updateReason
--]]
function Lootomatic.OnInventorySingleSlotUpdate(eventCode, bagId, slotId, isNewItem, itemSoundCategory, updateReason)
    if (not isNewItem) then return end
    Lootomatic.Log('onInventorySingleSlotUpdate', LootomaticLogger.DEBUG)
    local lootItem = LootomaticItem.LoadByBagAndSlot(bagId, slotId)
    --[[
    -- Check filters and mark item as junk if it matches
    -- a filter
    --]]
    for i,v in ipairs(Lootomatic.db.filters) do
        local filter = LootomaticFilter.New(v)
        if filter:IsMatch(lootItem) then
            Lootomatic.Log('Matched filter, marking item ' .. lootItem.name .. ' as Junk', LootomaticLogger.INFO)
            SetItemIsJunk(bagId, slotId, true)
            break
        end
    end
end

--[[
-- Display help
--]]
function LootomaticCommands.Help(cmd)
    Lootomatic.Log('config <key> <value>', LootomaticLogger.INFO)
    Lootomatic.Log('filters [list OR add OR delete OR show]', LootomaticLogger.INFO)
end

--[[
-- Manage configuration settings
--
-- @param string cmd
--]]
function LootomaticCommands.Config(cmd)
    if 'list' == cmd then
        d(Lootomatic.db.config)
        return
    end

    local options = { string.match(cmd,"^(%S*)%s*(.*)$") }

    if nil == options[1] or nil == options[2] then
        LootomaticCommands.Help()
        return
    end

    local config = Lootomatic.db.config

    if 'loglevel' == options[1] then
        newLogLevel = tonumber(options[2])
        -- @TODO check to make sure newLogLevel is valid value
        config.logLevel = newLogLevel
        Lootomatic.Log('Updated configuration', LootomaticLogger.INFO)
        return
    end

    if 'sellalljunk' == options[1] then
        config.sellAllJunk = toboolean(options[2])
        Lootomatic.Log('Updated configuration', LootomaticLogger.INFO)
        return
    end

    LootomaticCommands.Help()
end

--[[
-- What filter command to run
--
-- @param string cmd
--]]
function LootomaticCommands.Filters(cmd)
    if nil == cmd then
        LootomaticCommands.Help()
        return
    end

    -- /lootomatic filters list
    if 'list' == cmd then
        LootomaticCommands.FiltersList()
        return
    end

    -- /lootomatic filters clear
    if 'clear' == cmd then
        Lootomatic.db.filters = {}
        Lootomatic.Log('Filters have been cleared', LootomaticLogger.INFO)
        return
    end

    -- /lootomatic delete 1
    if string.match(cmd, '^delete%s%d+') then
        local i = string.match(cmd, '^delete%s(%d+)')
        local filter = Lootomatic.db.filters[tonumber(i)]
        if nil == filter then
            Lootomatic.Log('There is no filter on that index', LootomaticLogger.WARN)
            return
        end
        table.remove(Lootomatic.db.filters, i)
        Lootomatic.Log('Filter has been deleted, please run /lootomatic filters list for new indexes', LootomaticLogger.INFO)
        return
    end

    -- /lootomatic filters add name:Trash enabled:true condition.type:EqualTo condition.name:itemType condition.value:48
    if string.match(cmd, '^add%s.*') then
        local filter = {
            name = nil,
            enabled = true,
            rules = {
                {
                    condition = {
                        type = nil,
                        left = { name = nil, value = nil }
                    }
                }
            }
        }
        for k,v in string.gmatch(cmd, '%s([%w\.]+):?([%w]+)%s-') do
            -- Good place for a filter builder
            if 'name' == k then
                filter.name = v
            elseif 'enabled' == k then
                filter.enabled = toboolean(v)
            elseif 'condition.type' == k then
                filter.rules[1].condition.type = v
            elseif 'condition.name' == k then
                filter.rules[1].condition.left.name = v
            elseif 'condition.value' == k then
                filter.rules[1].condition.left.value = v
            end
        end
        table.insert(Lootomatic.db.filters, filter)
        Lootomatic.Log('Added new filter', LootomaticLogger.INFO)
        return
    end

    -- /lootomatic filters modify 1 name:Trash enabled:true condition.type:EqualTo condition.name:itemType condition.value:48
    if string.match(cmd, '^modify%s%d+%s.*') then
        local i = string.match(cmd, '^modify%s(%d+)')
        local filter = Lootomatic.db.filters[tonumber(i)]
        if nil == filter then
            Lootomatic.Log('There is no filter on that index', LootomaticLogger.WARN)
            return
        end
        for k,v in string.gmatch(cmd, '%s([%w\.]+):?([%w]+)%s-') do
            -- Good place for a filter builder
            if 'name' == k then
                filter.name = v
            elseif 'enabled' == k then
                filter.enabled = toboolean(v)
            elseif 'condition.type' == k then
                filter.rules[1].condition.type = v
            elseif 'condition.name' == k then
                filter.rules[1].condition.left.name = v
            elseif 'condition.value' == k then
                filter.rules[1].condition.left.value = v
            end
        end
        Lootomatic.Log('Modified filter', LootomaticLogger.INFO)
        return
    end

    -- /lootomatic filters show 1
    if string.match(cmd, '^show%s.*') then
        local i = string.match(cmd, '^show%s(%d+)')
        local filter = Lootomatic.db.filters[tonumber(i)]
        if nil == filter then
            Lootomatic.Log('There is no filter on that index', LootomaticLogger.WARN)
            return
        end
        filter = LootomaticFilter.New(filter)
        Lootomatic.Log('Name: ' .. filter:GetName(), LootomaticLogger.INFO)
        local enabled = 'No'
        if filter:IsEnabled() then
            enabled = 'Yes'
        end
        Lootomatic.Log('Enabled: ' .. enabled, LootomaticLogger.INFO)
        Lootomatic.Log('Rules: ', LootomaticLogger.INFO)
        for i,v in pairs(filter:GetRules()) do
            local c = v.condition
            Lootomatic.Log('['..i..'] ' .. c.left.name .. ' is ' .. c.type .. ' ' .. c.left.value, LootomaticLogger.INFO)
        end
        return
    end

    LootomaticCommands.Help()
end

--[[
-- List all known filters
--]]
function LootomaticCommands.FiltersList()
    for i,v in pairs(Lootomatic.db.filters) do
        local f = LootomaticFilter.New(v)
        Lootomatic.Log('[' .. i .. '] ' .. f:GetName(), LootomaticLogger.INFO)
    end
end

--[[
-- Used for parsing slash commands
--]]
function Lootomatic.Command(parameters)
    local options = {}
    local searchResult = { string.match(parameters,"^(%S*)%s*(.-)$") }
    for i,v in pairs(searchResult) do
        if (v ~= nil and v ~= "") then
            options[i] = string.lower(v)
        end
    end
    if #options == 0 or options[1] == "help" then
        LootomaticCommands.Help()
        return
    end

    if options[1] == 'config' then
        LootomaticCommands.Config(options[2])
        return
    end

    -- Want to manage filters
    if options[1] == 'filters' then
        LootomaticCommands.Filters(options[2])
        return
    end

    if options[1] == 'test' then
        local panel = LootomaticGUI:CreateTopLevelWindow('lootomaticgui')
        local a = LootomaticGUI:CreateControl('lootomatic_container', panel, CT_LABEL)
        local var1 = Ruler.Variable.New('Filter.itemType', 48)
        local var2 = Ruler.Variable.New('itemType')
        local o = Ruler.Operator.EqualTo.New(var1, var2)
        local r = Ruler.Rule.New(o)
        local context = {
            itemType = 48
        }
        local e = r:Evaluate(context)
        d(e)
        return
    end

    LootomaticCommands.Help()
end

--[[
-- @param integer eventCode
-- @param string  addOnName
--]]
function Lootomatic.OnAddOnLoaded(eventCode, addOnName)
    if (addOnName ~= Lootomatic.name) then
        return
    end
    Lootomatic.db = ZO_SavedVars:New('Lootomatic_Data', 1, nil, Lootomatic.defaults)

    -- Vendor events
    EVENT_MANAGER:RegisterForEvent(Lootomatic.name, EVENT_OPEN_STORE, Lootomatic.OnOpenStore)

    -- Inventory Events
    EVENT_MANAGER:RegisterForEvent(Lootomatic.name, EVENT_INVENTORY_SINGLE_SLOT_UPDATE, Lootomatic.OnInventorySingleSlotUpdate)

    -- Initialize slash command
    SLASH_COMMANDS['/lootomatic'] = Lootomatic.Command

    -- Initialize Settings
    Lootomatic_Settings.New(Lootomatic.db)
end

EVENT_MANAGER:RegisterForEvent(Lootomatic.name, EVENT_ADD_ON_LOADED, Lootomatic.OnAddOnLoaded)
